### ISSUE TITLE
Second gradient return None using tf.while_loop in TF2.x

### ISSUE THREAD

**Issue by KimiHsieh** [tensorflow/tensorflow#50278]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):Example provided below
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab
- TensorFlow installed from binary
- TensorFlow version (use command below): 2.5 and 1.15.2 both in Colab
- Python version: 3.7

**Describe the current behavior**
In tf2.5, when using `tf.while_loop`, we are able to call `.gradients()` in the `body()` of `tf.while_loop`. 
However, if we call `tf.GradientTape()`to get the first gradient outside the `tf.while_loop`, and then call `.gradients()` to get the second gradient in the `body()` of the `tf.while_loop`, the second gradient will be `None`.

**Standalone code to reproduce the issue**
[colab](https://colab.research.google.com/drive/1xV2vN_fZkkZz47AmoH-y-Fl1B6lmtpGB?usp=sharing)
```python3 
# step function
@tf.function
def step(dummy):
  with tf.GradientTape(persistent=True) as tape:
    loss = tf.reduce_sum(tf.pow(var, 2))  # loss = w^2
  grad = tape.gradient(loss, var)  # 1st gradient, grad = 2w
  train = loop(grad, var, dummy)

# Set loop funtion
def loop(grad, var, dummy):
  def cond(dummy):
    return True

  def body(dummy):
    s = tf.constant(1.)
    elemwise = math_ops.multiply(grad, array_ops.stop_gradient(s))
    grad_value = tf.gradients(elemwise, var) # 2nd gradient 
    # return None to grad_value, but expect grad_value is not None
    print('grad_value=', grad_value)
    return dummy

  return tf.while_loop(cond, body, [dummy])

# Set variables
var = tf.Variable(2., name= 'w1')
dummy = tf.constant(1.)
step(dummy)
```

**Describe the expected behavior**
Expect the grad_value is not `None`

**Other info / logs** 
 - tf2.5(**I focus on**) with `@tf.function` fails
 - tf1.15 In default graph mode works, but in eager mode with `@tf.function` fails
 - similar issue [this](https://github.com/tensorflow/tensorflow/issues/24866#issue-398513482) and [this](https://stackoverflow.com/questions/49555016/compute-gradients-for-each-time-step-of-tf-while-loop)

**Comment by tilakrayal** [comment]
@sachinprasadhs ,
I was able to reproduce the issue in tf v2.4,v2.5 and nightly.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/22b1834cffc0c92156e43d54bd157dfa/untitled50278.ipynb).

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tensorflow 2.5, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/1d4980812581f44f54ac5dcd3c8b7cb9/while_loop.ipynb). Thanks!

**Comment by kkimdev** [comment]
Although it's not officially deprecated (yet), `tf.gradients()` is TF1-style graph only API, could you try reformulating your code to use only `tf.GradientTape` ?

**Comment by KimiHsieh** [comment]
> Although it's not officially deprecated (yet), `tf.gradients()` is TF1-style graph only API, could you try reformulating your code to use only `tf.GradientTape` ?

Thanks for suggestion
I tried `tf.GradientTape` in the second gradient, it can't work neither(See the following code). 
My main concern is that why it can't work even 'tf.funtion()' puts everything into graph mode in TF2.X. Instead, in TF1.X,  it works also in graph mode. 

```python 
# step function
@tf.function
def step(dummy):
  with tf.GradientTape(persistent=True) as tape:
    loss = tf.reduce_sum(tf.pow(var, 2))  # loss = w^2
  grad = tape.gradient(loss, var)  # 1st gradient, grad = 2w
  train = loop(grad, var, dummy)

# Set loop funtion
def loop(grad, var, dummy):
  def cond(dummy):
    return True

  def body(dummy):
    s = tf.constant(1.)
    with tf.GradientTape(persistent=True) as tape:
      elemwise = math_ops.multiply(grad, array_ops.stop_gradient(s))
    grad_value = tape.gradient(elemwise, var) # 2nd gradient , grad_value= None
    # return None to grad_value, but expect grad_value is not None
    print('grad_value=', grad_value)
    return dummy

  return tf.while_loop(cond, body, [dummy])

# Set variables
var = tf.Variable(2., name= 'w1')
dummy = tf.constant(1.)
step(dummy)
```

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

By default, the resources held by a GradientTape are released as soon as GradientTape.gradient() method is called. To compute multiple gradients over the same computation, create a persistent gradient tape. This allows multiple calls to the gradient() method as resources are released when the tape object is garbage collected.
https://www.tensorflow.org/api_docs/python/tf/GradientTape

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
